gpbft/src/powertable.rs

```diff
@@ -77,6 +82,99 @@ impl FromIterator<PowerEntry> for PowerEntries {
     }
 }
 
+impl PowerEntries {
+    /// Scales power entries proportionally within the total power
+    pub fn scaled(&self) -> anyhow::Result<(Vec<i64>, i64)> {
+        // First pass: calculate total unscaled power
+        let mut total_unscaled = StoragePower::from(0);
+
+        for entry in self.iter() {
+            if entry.power <= NumZero::zero() {
+                return Err(anyhow!(
+                    "invalid non-positive power {} for participant {}",
+                    entry.power,
+                    entry.id
+                ));
+            }
+
+            total_unscaled += &entry.power;
+        }
+
+        // Second pass: scale each power and accumulate total
+        let mut scaled_powers = Vec::with_capacity(self.len());
+        let mut scaled_total = 0;
+
+        for entry in self.iter() {
+            let scaled_power = entry.scale_power(&total_unscaled)?;
+            scaled_powers.push(scaled_power);
+            scaled_total += scaled_power;
+        }
+
+        Ok((scaled_powers, scaled_total))
+    }
+
+    /// Serialize power table entries to CBOR format
+    pub fn serialize_cbor(&self) -> Vec<u8> {
+        use serde_cbor::Value;
+
+        let cbor_entries: Vec<Value> = self
+            .0
+            .iter()
+            .map(|entry| {
+                // Convert StoragePower to bytes like Go's big.Int.MarshalCBOR
+                // Go adds a leading zero byte to the big-endian representation
+                let power_bytes = {
+                    let mut bytes = entry.power.to_bytes_be().1;
+                    bytes.insert(0, 0);
+                    bytes
+                };
+
+                Value::Array(vec![
+                    Value::Integer(entry.id as i128),
+                    Value::Bytes(power_bytes),
+                    Value::Bytes(entry.pub_key.0.clone()),
+                ])
+            })
+            .collect();
+
+        let cbor_array = Value::Array(cbor_entries);
+        serde_cbor::to_vec(&cbor_array).unwrap()
+    }
+}
+
+/// Calculate the total power of a set of signers using their indices
+pub fn signer_scaled_total(scaled_powers: &[i64], signer_indices: &[u64]) -> anyhow::Result<i64> {
+    let mut signer_total = 0;
+    for &index in signer_indices {
+        let idx = index as usize;
+        if idx >= scaled_powers.len() {
+            return Err(anyhow!(
+                "signer index {} out of bounds (max: {})",
+                idx,
+                scaled_powers.len().saturating_sub(1)
+            ));
+        }
+        signer_total += scaled_powers[idx];
+    }
+
+    Ok(signer_total)
+}
+
+/// Check whether a portion of storage power is a strong quorum of the total
+pub fn is_strong_quorum(power: i64, power_total: i64) -> bool {
```

Comment:
Can the `power` and `power_total` be negative? If not, could we use `u64` and its stabilized [div_ceil](https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil)?

Reply:
Since it's the scaled value, it's actually bounded in `u16`.
But `go-f3` uses `i64`, so I followed that. https://github.com/filecoin-project/go-f3/blob/f9c95ae52d9ab64b4f673d31da4a7bcac610216e/gpbft/gpbft.go#L1489